Doc: Working with Logstash Plugins on ECK (#7702)
elastic/cloud-on-k8s: docs/orchestrating-elastic-stack-applications/logstash.asciidoc

[id="{p}-{page_id}"]
= Run Logstash on ECK

> experimental[]

Reviewer: Should we take this out now, or in a later PR?

* `NORMALIZED_CLUSTERNAME_ES_PASSWORD`
* `NORMALIZED_CLUSTERNAME_ES_SSL_CERTIFICATE_AUTHORITY`

where NORMALIZED_CLUSTERNAME is the value taken from the `clusterName` field of the `elasticsearchRef` property, capitalized, with `-` transformed to `_`. That is, `prod-es` would become `PROD_ES`.

[NOTE]
--
* The `clusterName` value should be unique across all referenced {es} instances in the same {ls} spec.
* The {ls} ECK operator creates a user called `eck_logstash_user_role` when an `elasticsearchRef` is specified. This user has the following permissions:
+
```

>   "cluster": ["monitor", "manage_ilm", "read_ilm", "manage_logstash_pipelines", "manage_index_templates", "cluster:admin/ingest/pipeline/get",],

Author: ```suggestion
  "cluster": ["monitor", "manage_ilm", "read_ilm", "manage_logstash_pipelines", "manage_index_templates", "cluster:admin/ingest/pipeline/get",]
```

>     }
> 

Author: ```suggestion
    }
]
```

<4> The CA certificate in PEM format to secure communication to the Elasticsearch cluster (optional).
<5> The `secretName` and `name` attributes are mutually exclusive, you have to choose one or the other.

> TIP: Always specify the port in the URL when {LS} is connecting to an external Elasticsearch Cluster.

Author: There are a LOT of these.  I'll scrub these in a follow-up PR.

> TIP: Always specify the port in the URL when {LS} is connecting to an external Elasticsearch Cluster.

Author: There are lots of these instances. I will clean up remaining in a subsequent PR.

[source,yaml]
----
spec:
  podTemplate:
    spec:
      initContainers:
      - name: download-postgres
        command: ["/bin/sh"]
        args: ["-c", "curl -o /data/postgresql.jar -L https://jdbc.postgresql.org/download/postgresql-42.6.0.jar"]
        volumeMounts:
          - name: workdir
            mountPath: /data
      containers:
        - name: logstash
          volumeMounts:
            - name: workdir
              mountPath: /usr/share/logstash/jars <1>
  pipelines:
    - pipeline.id: main
      config.string: |
        input {
          jdbc {
             jdbc_driver_library => "/usr/share/logstash/jars/postgresql.jar"
             jdbc_driver_class => "org.postgresql.Driver"
             <2>
          }
        }
----
<1> Should match the `mountPath` of the `container`
<2> Remainder of plugin configuration goes here

[id="{p}-logstash-custom-images"]
===== Adding files using a custom Docker image

This example downloads the same `postgres` JDBC driver, and adds it to the {ls} classpath in the Docker image.

First, create a Dockerfile based on the {ls} Docker image.
Download the JDBC driver, and save it alongside the other JAR files in the {ls} classpath:

> From docker.elastic.co/logstash/logstash:{version}, run:
> [subs="attributes"]

Author: ```suggestion
From docker.elastic.co/logstash/logstash:{version}, run:

["source","shell",subs="attributes"]
```

[id="{p}-logstash-inputs-external-state"]
==== Input plugins: external source stores state

{ls} installations that use input plugins that retrieve data from an external source, and **rely on the external source to store state** can scale based on the parameters of the external source.

For example, a {ls} installation that uses a {logstash-ref}/plugins-inputs-kafka.html[`logstash-input-kafka`] plugin to retrieve data can scale the number of pods up to the number of partitions used, as a partition can have at most one consumer belonging to the same consumer group. 
Any pods created beyond that threshold cannot be scheduled to receive data.

Examples of these plugins include {logstash-ref}/plugins-inputs-kafka.html[`logstash-input-kafka`], {logstash-ref}/plugins-inputs-azure_event_hubs.html[`logstash-input-azure_event_hubs`], and {logstash-ref}/plugins-inputs-kinesis.html[`logstash-input-kinesis`].

[id="{p}-logstash-working-with-plugin-considerations"]

> === Plugin-specific considerations

Author: I added it.  Note that I also reformatted the section to use headings instead of the definition list format. Please check to see if you're OK with the description I used before your example, or if you would like for me to tweak it.

> === Plugin-specific considerations

Author: Please see if I implemented this as you pictured it.

> === Plugin-specific considerations

Reviewer 2: btw, the source code in the [original note](https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-logstash-configuration.html#k8s-logstash-pipelines-es) is missing `]` at the end and having an extra `,` after `"cluster:admin/ingest/pipeline/get"` which will cause syntax error :)

This is the correct yaml
```yaml
"cluster": ["monitor", "manage_ilm", "read_ilm", "manage_logstash_pipelines", "manage_index_templates", "cluster:admin/ingest/pipeline/get"],
"indices": [
  {
    "names": [ "logstash", "logstash-*", "ecs-logstash", "ecs-logstash-*", "logs-*", "metrics-*", "synthetics-*", "traces-*" ],
    "privileges": ["manage", "write", "create_index", "read", "view_index_metadata"]
  }
]
```

Some plugins have additional requirements and guidelines for optimal performance in a {ls} ECK environment.

TIP: Use these guidelines _in addition_ to the general guidelines provided in <<{p}-logstash-working-with-plugins-scaling>>.

{logstash-ref}/plugins-integrations-logstash.html[Logstash integration plugin]::
When your pipeline uses `logstash-integration-logstash`, add `keepalive=>false` to the {logstash-ref}/plugins-outputs-logstash.html[logstash-output] definition to ensure that load balancing works correctly rather than keeping affinity to the same pod.

{logstash-ref}/plugins-filters-elastic_integration.html[Elastic_integration filter plugin]::

> The elastic_integration filter allows the use of `ElasticsearchRef` and environment variables.

Author: To make this work, I added a new heading called "elasticsearchRefs for establishing a secured connection". I'm linking to the new section from the elastic_integration filter section.  

@karenzone note to self:  Restructure to make this flow better in the next iteration

          disabled: true
----

[id="{p}-logstash-plugins"]
== {ls} plugins

The power of {ls} is in the plugins--{logstash-ref}/input-plugins.html[inputs], {logstash-ref}/output-plugins.html[outputs], {logstash-ref}/filter-plugins.html[filters,] and {logstash-ref}/codec-plugins.html[codecs].

In {ls} on ECK, you can use the same plugins that you use for other {ls} instances--including Elastic-supported, community-supported, and custom plugins.
However, you may have other factors to consider, such as how you configure your {k8s} resources, how you specify additional resources, and how you scale your {ls} installation.  

In this section, we'll cover: 

> * Providing additional resources for plugins (read-only and writable storage)

Author: LOL. You read my mind. That's the question I was playing with here.  :-D 

I've had time to formulate my thinking on this.  When there's a list immediately followed by the content we're linking to, I'll use a link in _all_ list items for consistency and parallelism.  

BUT in cases where we have a single link to the content immediately after the current content, I'll continue recommending that we don't add a link.  It's weird when you click a link and realize you didn't go anywhere. :Thank you for helping me firm up my thinking on this.  :-)

Many plugins require or allow read-only assets in order to work correctly.
These may be ConfigMaps or Secrets files that have a 1 MB limit, or larger assets such as JDBC drivers, that need to be stored in a PersistentVolume.

[id="{p}-logstash-working-with-plugins-small-ro"]
===== ConfigMaps and Secrets (1 MB max)

Each instance of a `ConfigMap` or `Secret` has a maximum size of `1 MB`. 
For larger read-only assets, check out <<{p}-logstash-working-with-plugins-large-ro>>.

In the plugin documentation, look for configurations that call for a `path` or an `array` of `paths`. 

**Sensitive assets, such as private keys**

Some plugins need access to private keys or certificates in order to access an external resource.
Make the keys or certificates available to the {ls} resource in your manifest.

TIP: These settings are typically identified by an `ssl_` prefix, such as `ssl_key`, `ssl_keystore_path`, `ssl_certificate`, for example.

To use these in your manifest, create a Secret representing the asset, a Volume in your `podTemplate.spec` containing that Secret, and then mount that Volume with a VolumeMount in the `podTemplateSpec.container` section of your {ls} resource.

> 

Author: They're definitely more discoverable this way.  If we collapse them, they would be easy to miss if a user scans the topic.  I'm inclined to leave them expanded and obvious.  If you have strong feelings to the contrary, I will make them collapsible.

> 

Reviewer: I guess my concern is that the examples might be geared towards a K8s beginner, and might be noise to a more experienced user

> 

Author: ALL  code samples?

          }
        }
        filter {
          ruby {
            path => "/usr/share/logstash/data/drop_percentage.rb"
            script_params => { "percentage" => 0.9 }
          }
        }
----

[id="{p}-logstash-working-with-plugins-large-ro"]
==== Larger read-only assets (1 MB+)

Some plugins require or allow access to static read-only files that exceed the `1 MB` limit imposed by ConfigMap and Secret. 
For example, you may need JAR files to load drivers when using a JDBC or JMS plugin, or a large {logstash-ref}/plugins-filters-translate.html[`logstash-filter-translate`] dictionary.

Add these files using one or more of these options:

* **PersistentVolume.** Create a PersistentVolume that points to the static content, and create a PersistentVolumeClaim to request storage to be used by the {ls} resource.
* **initContainer.** Use an `initContainer` to add the files before the main container starts. This approach may be appropriate when the static content is hosted remotely.
* **Custom docker image.** 

Check out <<{p}-bundles-plugins>> for more details on which option might be most suitable for you.

[id="{p}-logstash-pv"]
===== Adding files using a PersistentVolume

> 

Author: ToDo: Add content for PersistentVolume

Author: Done. Thanks, Rob.

        volumeMounts:
          - name: workdir
            mountPath: /data
      containers:
        - name: logstash
          volumeMounts:
            - name: workdir
              mountPath: /usr/share/logstash/jars <1>
  pipelines:
    - pipeline.id: main
      config.string: |
        input {
          jdbc {
             jdbc_driver_library => "/usr/share/logstash/jars/postgresql.jar"
             jdbc_driver_class => "org.postgresql.Driver"
             <2>
          }
        }
----
<1> Should match the `mountPath` of the `container`
<2> Remainder of plugin configuration goes here

[id="{p}-logstash-custom-images"]
===== Adding files using a custom Docker image

This example downloads the same `postgres` JDBC driver, and adds it to the {ls} classpath in the Docker image.

First, create a Dockerfile based on the {ls} Docker image.
Download the JDBC driver, and save it alongside the other JAR files in the {ls} classpath:

From docker.elastic.co/logstash/logstash:{version}, run:
[source,js]
----
curl -o /usr/share/logstash/logstash-core/lib/jars/postgresql.jar -L https://jdbc.postgresql.org/download/postgresql-42.6.0.jar <1>

> ----

Author: I wasn't sure if you wanted the "From docker.elastic.co/logstash/logstash:{version}, run:" to come out or not.  I left it for now, but we can delete it if you'd like.

> ----

Reviewer: I want it in the `----` block as a code sample, not as an explanation

* The ability of a {ls} installation to scale horizontally is bound by its most restrictive plugin(s). Even if all pipelines are using {logstash-ref}/plugins-inputs-elastic_agent.html[`logstash-input-elastic_agent`] or {logstash-ref}/plugins-inputs-beats.html[`logstash-input-beats`] which should enable full horizontal scaling, introducing a more restrictive input or filter plugin forces the restrictions for pod scaling associated with that plugin.
****

[id="{p}-logstash-agg-filters"]
==== Filter plugins: aggregating filters

{ls} installations that use aggregating filters should be treated with particular care: 

* They *must* specify `pipeline.workers=1` for any pipelines that use them.
* The number of pods cannot be scaled above 1.

// Note to authors and reviewers: Is `logstash-filter-*` a real filter???

Examples of aggregating filters include {logstash-ref}/plugins-filters-aggregate.html[`logstash-filter-aggregate`], `logstash-filter-*`, and any {logstash-ref}/plugins-filters-ruby.html[`logstash-filter-ruby`] implementations that perform aggregations.

[id="{p}-logstash-inputs-data-pushed"]

> ==== Input plugins: data pushed to the input

Reviewer: What do you think about as a comparison between the push and pull model:

"events pushed to Logstash" vs "Logstash pulls events"

or "Logstash receives events" vs "Logstash retrieves events"

and maybe a quick explainer of push vs. pull model

Author: I like that approach!  That's what I was initially thinking, but wasn't confident enough in my understanding to put it in writing. :-)

Author: I tried to work both "pushed to LS" and "LS receives events" into the heading and text, restating for additional context. Please see what you think.

[id="{p}-logstash-inputs-external-state"]
==== Input plugins: external source stores state

{ls} installations that use input plugins that retrieve data from an external source, and **rely on the external source to store state** can scale based on the parameters of the external source.

For example, a {ls} installation that uses a {logstash-ref}/plugins-inputs-kafka.html[`logstash-input-kafka`] plugin to retrieve data can scale the number of pods up to the number of partitions used, as a partition can have at most one consumer belonging to the same consumer group. 
Any pods created beyond that threshold cannot be scheduled to receive data.

Examples of these plugins include {logstash-ref}/plugins-inputs-kafka.html[`logstash-input-kafka`], {logstash-ref}/plugins-inputs-azure_event_hubs.html[`logstash-input-azure_event_hubs`], and {logstash-ref}/plugins-inputs-kinesis.html[`logstash-input-kinesis`].

[id="{p}-logstash-working-with-plugin-considerations"]
=== Plugin-specific considerations

Some plugins have additional requirements and guidelines for optimal performance in a {ls} ECK environment.

> // Note to authors and reviewers: right now, there doesn't appear to be any overlap. Should we keep this tip?

Reviewer: I like it

            username => "${ECK_ES_USER}"
            password => "${ECK_ES_PASSWORD}"
            ssl_certificate_authorities => "${ECK_ES_SSL_CERTIFICATE_AUTHORITY}"
          }

----
+
The Elastic_integration filter requires certain roles to be configured on the {es} cluster to enable {ls} to read ingest pipelines.
+
[source, yaml,subs="attributes,+macros,callouts"]
----
//Sample role definition
kind: Secret
apiVersion: v1
metadata:
  name: my-roles-secret
stringData:
  roles.yml: |-
    eck_logstash_user_role:
      cluster: [ "monitor", "manage_index_templates", "read_pipeline"]
----

{logstash-ref}/plugins-inputs-elastic_agent.html[Elastic_agent input] and {logstash-ref}/plugins-inputs-beats.html[Beats input] plugins::
Set the `ttl` value on Elastic Agent or Beat to ensure that load is distributed appropriately.

[id="{p}-logstash-working-with-custom-plugins"]
=== Adding custom plugins

If you need plugins in addition to those included in the standard {ls} distribution, you can add them. 
Create a custom Docker image that includes the installed plugins, using the `bin/logstash-plugin install` utility to add more plugins to the image so that they can be used by {ls} pods.

> // NOTE TO AUTHORS/REVIEWERS: `logstash-filter-elastic_integration` is a default plugin as of 8.13.0. Sub out and replace with another good example? 

Reviewer: Yeah, let's take it out

> This sample Dockerfile installs the {logstash-ref}/plugins-filters-tld.html[`logstash-filter-tld`] and {logstash-ref}/plugins-filters-elastic_integration.html[`logstash-filter-elastic_integration`] plugins to the official {ls} Docker image:

Reviewer: ```suggestion
This sample Dockerfile installs the {logstash-ref}/plugins-filters-tld.html[`logstash-filter-tld`] plugin to the official {ls} Docker image:
```

["source","shell",subs="attributes"]
----
FROM docker.elastic.co/logstash/logstash:{version}

> RUN bin/logstash-plugin install logstash-filter-tld logstash-filter-elastic_integration

Reviewer: ```suggestion
RUN bin/logstash-plugin install logstash-filter-tld
```

[id="{p}-logstash-working-with-plugin-considerations"]
=== Plugin-specific considerations

Some plugins have additional requirements and guidelines for optimal performance in a {ls} ECK environment.

* <<{p}-logstash-plugin-considerations-ls-integration>>
* <<{p}-logstash-plugin-considerations-es-output>>
* <<{p}-logstash-plugin-considerations-integration-filter>>
* <<{p}-logstash-plugin-considerations-agent-beats>>

TIP: Use these guidelines _in addition_ to the general guidelines provided in <<{p}-logstash-working-with-plugins-scaling>>.

[id="{p}-logstash-plugin-considerations-ls-integration"]
==== {ls} integration plugin 

When your pipeline uses the {logstash-ref}/plugins-integrations-logstash.html[`Logstash integration`] plugin, add `keepalive=>false` to the {logstash-ref}/plugins-outputs-logstash.html[logstash-output] definition to ensure that load balancing works correctly rather than keeping affinity to the same pod.

[id="{p}-logstash-plugin-considerations-es-output"]
==== Elasticsearch output plugin

> The {logstash-ref}/plugins-output-elasticsearch.html[`elasticsearch output`] plugin requires certain roles to be configured in order to enable {ls} to communicate with {es}.

Author: ```suggestion
The {logstash-ref}/plugins-outputs-elasticsearch.html[`elasticsearch output`] plugin requires certain roles to be configured in order to enable {ls} to communicate with {es}.
```

> [source, logstash]
> ----
> "cluster": ["monitor", "manage_ilm", "read_ilm", "manage_logstash_pipelines", "manage_index_templates", "cluster:admin/ingest/pipeline/get",],
>   "indices": [
>     {
>       "names": [ "logstash", "logstash-*", "ecs-logstash", "ecs-logstash-*", "logs-*", "metrics-*", "synthetics-*", "traces-*" ],
>       "privileges": ["manage", "write", "create_index", "read", "view_index_metadata"]
>     }
> ----

Reviewer 2: ```suggestion
You can customize roles in {es}. Check out <<{p}-users-and-roles,creating custom roles>>

[source, logstash]
----
kind: Secret
apiVersion: v1
metadata:
  name: my-roles-secret
stringData:
  roles.yml: |-
    eck_logstash_user_role:
      "cluster": ["monitor", "manage_ilm", "read_ilm", "manage_logstash_pipelines", "manage_index_templates", "cluster:admin/ingest/pipeline/get",],
      "indices": [
        {
          "names": [ "logstash", "logstash-*", "ecs-logstash", "ecs-logstash-*", "logs-*", "metrics-*", "synthetics-*", "traces-*" ],
          "privileges": ["manage", "write", "create_index", "read", "view_index_metadata"]
        }
      ]
----
```

Reviewer 2: I added a pointer to https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-users-and-roles.html#k8s_creating_custom_roles which has the necessary steps to customzie roles. 
It is not a simple step just k apply -f resources.yaml. It requires configuration in ES ECK

Author: ^^ I'm not seeing the pointer. Where did you add it?

The {logstash-ref}/plugins-filters-elastic_integration.html[`elastic_integration filter`] plugin allows the use of <<k8s-logstash-esref,`ElasticsearchRef`>> and environment variables.

[source, logstash]
----
  elastic_integration {
            pipeline_name => "logstash-pipeline"
            hosts => [ "${ECK_ES_HOSTS}" ]
            username => "${ECK_ES_USER}"
            password => "${ECK_ES_PASSWORD}"
            ssl_certificate_authorities => "${ECK_ES_SSL_CERTIFICATE_AUTHORITY}"
          }

----

The Elastic_integration filter requires certain roles to be configured on the {es} cluster to enable {ls} to read ingest pipelines.

[source, yaml,subs="attributes,+macros,callouts"]
----
# Sample role definition
kind: Secret
apiVersion: v1
metadata:
  name: my-roles-secret
stringData:
  roles.yml: |-
    eck_logstash_user_role:
      cluster: [ "monitor", "manage_index_templates", "read_pipeline"]
----

[id="{p}-logstash-plugin-considerations-agent-beats"]
==== Elastic_agent input and Beats input plugins

When you use the {logstash-ref}/plugins-inputs-elastic_agent.html[Elastic_agent input] or the {logstash-ref}/plugins-inputs-beats.html[Beats input],

> set the {logstash-ref}/plugins-inputs-beats.html#_ttl[`ttl`] value on the input to ensure that load is distributed appropriately.

Author: ```suggestion
set the {filebeat-ref}logstash-output.html#_ttl[`ttl`] value on the Agent or Beat to ensure that load is distributed appropriately.
```

For larger read-only assets, check out <<{p}-logstash-working-with-plugins-large-ro>>.

In the plugin documentation, look for configurations that call for a `path` or an `array` of `paths`. 

**Sensitive assets, such as private keys**

Some plugins need access to private keys or certificates in order to access an external resource.
Make the keys or certificates available to the {ls} resource in your manifest.

TIP: These settings are typically identified by an `ssl_` prefix, such as `ssl_key`, `ssl_keystore_path`, `ssl_certificate`, for example.

To use these in your manifest, create a Secret representing the asset, a Volume in your `podTemplate.spec` containing that Secret, and then mount that Volume with a VolumeMount in the `podTemplateSpec.container` section of your {ls} resource.

First, create your secrets.

[source,bash]
----
kubectl create secret generic logstash-crt --from-file=logstash.crt
kubectl create secret generic logstash-key --from-file=logstash.key
----

> Then, create your Logstash CRD.

Reviewer 3: ```suggestion
Then, create your Logstash resource.
```
or `object` or `manifest`. CRD is the custom resource definition that extends the k8s API and makes it possible to create Logstash objects in the first place, so to speak the defintion of this "definition".

You can use these for a variety of reasons. 
Examples include adding custom `grok` patterns for {logstash-ref}/plugins-filters-grok.html[`logstash-filter-grok`] to use for lookup, source code for [`logstash-filter-ruby`], a dictionary for {logstash-ref}/plugins-filters-translate.html[`logstash-filter-translate`] or the location of a SQL statement for {logstash-ref}/plugins-inputs-jdbc.html[`logstash-input-jdbc`]. 
Make these files available to the {ls} resource in your manifest.

TIP: In the plugin documentation, these plugin settings are typically identified by `path` or an `array` of `paths`.

To use these in your manifest, create a ConfigMap or Secret representing the asset, a Volume in your `podTemplate.spec` containing the ConfigMap or Secret, and mount that Volume with a VolumeMount in your `podTemplateSpec.container` section of your {ls} resource.

This example illustrates configuring a ConfigMap from a ruby source file, and including it in a {logstash-ref}/plugins-filters-ruby.html[`logstash-filter-ruby`] plugin.

First, create the ConfigMap.

[source,bash]
----
kubectl create configmap ruby --from-file=drop_some.rb
----

> Then, create your Logstash CRD.

Reviewer 3: ```suggestion
Then, create your Logstash resource.
```

             jdbc_driver_class => "org.postgresql.Driver"
             last_metadata_path => "/usr/share/logstash/data/main/logstash_jdbc_last_run <1>
          }
        }
----
<1> If you are using more than one plugin of the same type, specify a unique location for each plugin to use.

If the default `logstash-data` volume is insufficient for your needs, see the volume section for details on how to add additional volumes.

[id="{p}-logstash-working-with-plugins-writable-temp"]
===== Writable staging or temporary data

Some {ls} plugins write data to a staging directory or file before processing for input, or outputting to their final destination. 
Often these staging folders can be persisted across restarts to avoid duplicating processing of data.

In the plugin documentation, look for  names such as `tmp_directory`, `temporary_directory`, `staging_directory`.

To persist data across pod restarts, set this value to point to the default `logstash-data` volume or your own PersistentVolumeClaim. 

[source, yaml,subs="attributes,+macros,callouts"]
----
spec:
  pipelines:
    - pipeline.id: main
      config.string: |
        output {
          s3 {

>              id => 'main_s3_output'

Author: ```suggestion
             id => "main_s3_output"
```

Author: @robbavey GitHub didn't know how to resolve when I tried to apply two suggestions to the same line, and my ability to apply your suggestion went away. That's why I'm reintroducing your previous comment.

          "privileges": ["manage", "write", "create_index", "read", "view_index_metadata"]
        }
      ]
----

[id="{p}-logstash-plugin-considerations-integration-filter"]
==== Elastic_integration filter plugin

The {logstash-ref}/plugins-filters-elastic_integration.html[`elastic_integration filter`] plugin allows the use of <<k8s-logstash-esref,`ElasticsearchRef`>> and environment variables.

[source, logstash]
----
  elastic_integration {
            pipeline_name => "logstash-pipeline"
            hosts => [ "${ECK_ES_HOSTS}" ]
            username => "${ECK_ES_USER}"
            password => "${ECK_ES_PASSWORD}"
            ssl_certificate_authorities => "${ECK_ES_SSL_CERTIFICATE_AUTHORITY}"
          }

----

The Elastic_integration filter requires certain roles to be configured on the {es} cluster to enable {ls} to read ingest pipelines.

[source, yaml,subs="attributes,+macros,callouts"]
----
# Sample role definition
kind: Secret
apiVersion: v1
metadata:
  name: my-roles-secret
stringData:
  roles.yml: |-
    eck_logstash_user_role:
      cluster: [ "monitor", "manage_index_templates", "read_pipeline"]
----

[id="{p}-logstash-plugin-considerations-agent-beats"]

> ==== Elastic_agent input and Beats input plugins

Reviewer 3: ```suggestion
==== Elastic Agent input and Beats input plugins
```

> When you use the {logstash-ref}/plugins-inputs-elastic_agent.html[Elastic_agent input] or the {logstash-ref}/plugins-inputs-beats.html[Beats input],

Reviewer 3: ```suggestion
When you use the {logstash-ref}/plugins-inputs-elastic_agent.html[Elastic Agent input] or the {logstash-ref}/plugins-inputs-beats.html[Beats input],
```

        volumeMounts:
          - name: workdir
            mountPath: /data
      containers:
        - name: logstash
          volumeMounts:
            - name: workdir
              mountPath: /usr/share/logstash/jars <1>
  volumeClaimTemplates:
    - metadata:
        name: workdir
      spec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 50Mi
  pipelines:
    - pipeline.id: main
      config.string: |
        input {
          jdbc {
             jdbc_driver_library => "/usr/share/logstash/jars/postgresql.jar"
             jdbc_driver_class => "org.postgresql.Driver"
             <2>
          }
        }
----
<1> Should match the `mountPath` of the `container`
<2> Remainder of plugin configuration goes here

[id="{p}-logstash-custom-images"]
===== Add files using a custom Docker image

This example downloads the same `postgres` JDBC driver, and adds it to the {ls} classpath in the Docker image.

First, create a Dockerfile based on the {ls} Docker image.
Download the JDBC driver, and save it alongside the other JAR files in the {ls} classpath:

> From docker.elastic.co/logstash/logstash:{version}, run:

Reviewer: ```suggestion

```

Reviewer: This should only be in the code snippet

             jdbc_driver_class => "org.postgresql.Driver"
             last_metadata_path => "/usr/share/logstash/data/main/logstash_jdbc_last_run <1>
          }
        }
----
<1> If you are using more than one plugin of the same type, specify a unique location for each plugin to use.

If the default `logstash-data` volume is insufficient for your needs, see the volume section for details on how to add additional volumes.

[id="{p}-logstash-working-with-plugins-writable-temp"]
===== Writable staging or temporary data

Some {ls} plugins write data to a staging directory or file before processing for input, or outputting to their final destination. 
Often these staging folders can be persisted across restarts to avoid duplicating processing of data.

In the plugin documentation, look for  names such as `tmp_directory`, `temporary_directory`, `staging_directory`.

To persist data across pod restarts, set this value to point to the default `logstash-data` volume or your own PersistentVolumeClaim. 

[source, yaml,subs="attributes,+macros,callouts"]
----
spec:
  pipelines:
    - pipeline.id: main
      config.string: |
        output {
          s3 {

>              id => 'main_s3_output`

Reviewer: ```suggestion
             id => 'main_s3_output'
```

>              id => 'main_s3_output`

Reviewer: ```suggestion
             id => "main_s3_output"
```

Author: Looks like this might be a duplicate, but I'm going to try to apply it to see what happens.

Author: Actually not a duplicate, but two conflicting changes to Line 1139.  I tried to apply both, and it appears the the first one is the one that was applied.  Which is correct?

Reviewer: I would prefer to use ` "main_s3_output"`, to keep the quoting consistent across sample code

This example illustrates configuring a ConfigMap from a ruby source file, and including it in a {logstash-ref}/plugins-filters-ruby.html[`logstash-filter-ruby`] plugin.

First, create the ConfigMap.

[source,bash]
----
kubectl create configmap ruby --from-file=drop_some.rb
----

Then, create your Logstash CRD.

[source,yaml]
----
spec:
  podTemplate:
    spec:
      volumes:
        - name: ruby_drop
          configMap:
            name: ruby
      containers:
        - name: logstash
          volumeMounts:
            - name: ruby_drop
              mountPath: "/usr/share/logstash/data/drop_percentage.rb"
              readOnly: true
  pipelines:
    - pipeline.id: main
      config.string: |
        input {
          beats {
            port => 5044
          }
        }
        filter {
          ruby {
            path => "/usr/share/logstash/data/drop_percentage.rb"
            script_params => { "percentage" => 0.9 }
          }
        }
----

[id="{p}-logstash-working-with-plugins-large-ro"]

> ==== Larger read-only assets (1 MiB+)
> 
> Some plugins require or allow access to static read-only files that exceed the 1 MiB (mebibyte) limit imposed by ConfigMap and Secret. 
> For example, you may need JAR files to load drivers when using a JDBC or JMS plugin, or a large {logstash-ref}/plugins-filters-translate.html[`logstash-filter-translate`] dictionary.
> 
> Add these files using one or more of these options:
> 
> * **PersistentVolume.** Create a PersistentVolume that points to the static content, and create a PersistentVolumeClaim to request storage to be used by the {ls} resource.
> * **initContainer.** Use an `initContainer` to add the files before the main container starts. This approach may be appropriate when the static content is hosted remotely.
> * **Custom docker image.** 
> 
> Check out <<{p}-bundles-plugins>> for more details on which option might be most suitable for you.
> 
> [id="{p}-logstash-pv"]
> ===== Adding files using a PersistentVolume
> 
> [id="{p}-logstash-ic"]
> ===== Adding files using an initContainer
> 
> This example creates an `initContainer` to download a PostgreSQL JDBC driver JAR file, places it in a volume mount accessible to the main `container`, and then uses it in a JDBC input in the pipeline configuration.
> 

Reviewer: ```suggestion
==== Larger read-only assets (1 MiB+)

Some plugins require or allow access to static read-only files that exceed the 1 MiB (mebibyte) limit imposed by ConfigMap and Secret.
For example, you may need JAR files to load drivers when using a JDBC or JMS plugin, or a large {logstash-ref}/plugins-filters-translate.html[`logstash-filter-translate`] dictionary.

Add these files using one or more of these options:

* **PersistentVolume.** Add a volumeClaimTemplate and a volumeMount to your {ls} resource and upload data to that volume, either using an `initContainer`, or direct upload if your Kubernetes provider supports it. You can use the default `logstash-data` volumeClaimTemplate , or a custom one depending on your storage needs.
* **Custom docker image.** Use a custom docker image that includes the static content that your Logstash pods will need.

Check out <<{p}-bundles-plugins>> for more details on which option might be most suitable for you.

[id="{p}-logstash-ic"]
===== Adding files using an initContainer

This example creates a volumeClaimTemplate called `workdir`, with volumeMounts referring to this mounted to the main container and an initContainer. The initContainer initiates a  download of a PostgreSQL JDBC driver JAR file, and stored it the volumeMount, which is then used in the JDBC input in the pipeline configuration.

```

The {logstash-ref}/plugins-filters-elastic_integration.html[`elastic_integration filter`] plugin allows the use of <<k8s-logstash-esref,`ElasticsearchRef`>> and environment variables.

[source, logstash]
----
  elastic_integration {
            pipeline_name => "logstash-pipeline"
            hosts => [ "${ECK_ES_HOSTS}" ]
            username => "${ECK_ES_USER}"
            password => "${ECK_ES_PASSWORD}"
            ssl_certificate_authorities => "${ECK_ES_SSL_CERTIFICATE_AUTHORITY}"
          }

----

The Elastic_integration filter requires certain roles to be configured on the {es} cluster to enable {ls} to read ingest pipelines.

[source, yaml,subs="attributes,+macros,callouts"]
----
# Sample role definition
kind: Secret
apiVersion: v1
metadata:
  name: my-roles-secret
stringData:
  roles.yml: |-
    eck_logstash_user_role:
      cluster: [ "monitor", "manage_index_templates", "read_pipeline"]
----

[id="{p}-logstash-plugin-considerations-agent-beats"]
==== Elastic_agent input and Beats input plugins

When you use the {logstash-ref}/plugins-inputs-elastic_agent.html[Elastic_agent input] or the {logstash-ref}/plugins-inputs-beats.html[Beats input],

> set the {filebeat-ref}logstash-output.html#_ttl[`ttl`] value on the Agent or Beat to ensure that load is distributed appropriately.

Author: ```suggestion
set the {filebeat-ref}/logstash-output.html#_ttl[`ttl`] value on the Agent or Beat to ensure that load is distributed appropriately.
```
